examples/vision/fixres.py

```diff
@@ -0,0 +1,319 @@
+"""
+Title: FixRes: Fixing train-test resolution discrepency
+Author: [Sayak Paul](https://twitter.com/RisingSayak)
+Date created: 2021/10/08
+Last modified: 2021/10/08
+Description: Mitigating resolution discrepency between training and test sets.
+"""
+"""
+## Introduction 
+
+It is a common practice to use the same input image resolution while training vision
+models. However, as investigated in
+[Fixing the train-test resolution discrepancy](https://arxiv.org/abs/1906.06423)
+(Touvron et al.), this practice leads to suboptimal performance. Data augmentation
+is an inseparable component for training deep neural networks. For vision models, we
+typically use random resized crops during training and center crops during inference.
+This introduces a discrepency in the object sizes seen during training and inference.
+As shown by Touvron et al., if we can fix this discrepency then we can significantly
+boost model performance. 
+
+In this example, we will implement the techniques introduced by Touvron et al. in this
+regard. As a note, these techniques are collectively referred to as **FixRes**.
+"""
+
+"""
+## Imports
+"""
+
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow as tf
+
+import tensorflow_datasets as tfds
+
+tfds.disable_progress_bar()
+
+import matplotlib.pyplot as plt
+
+"""
+## Load the `tf_flowers` dataset
+"""
+
+train_dataset, val_dataset = tfds.load(
+    "tf_flowers", split=["train[:90%]", "train[90%:]"], as_supervised=True
+)
+
+num_train = tf.data.experimental.cardinality(train_dataset)
+num_val = tf.data.experimental.cardinality(val_dataset)
+print(f"Number of training examples: {num_train}")
+print(f"Number of validation examples: {num_val}")
+
+"""
+## Data preprocessing utilities
+"""
+
+"""
+We will three datasets:
+
+1. A dataset with a smaller resolution - 128x128.
+2. Two dataset with a larger resolution - 224x224.
+
+Datasets with a larger resolution will differ in terms of the augmentation transforms.
+
+The idea of FixRes is to first train a model on a smaller resolution and then fine-tune
+it on the larger one. This simple yet effective recipe leads to non-trivial performance
+improvements. Please refer to the [original paper](https://arxiv.org/abs/1906.06423) for
+results.
+"""
+
+# Reference: https://github.com/facebookresearch/FixRes/blob/main/transforms_v2.py.
+
+batch_size = 128
+auto = tf.data.AUTOTUNE
+smaller_size = 128
+bigger_size = 224
+
+size_for_resizing = int((bigger_size / smaller_size) * bigger_size)
+central_crop_layer = layers.CenterCrop(bigger_size, bigger_size)
+
+
+def make_dataset(
+    dataset: tf.data.Dataset,
+    train: bool,
+    image_size: int = smaller_size,
+    fixres: bool = True,
+):
+    def preprocess_initial(image, label):
```

Comment:
Move the 2 utility functions outside of `make_dataset` (alternatively, you could use 3 different methods on a class...)